dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/EPersonRestRepository.java

```diff
@@ -82,8 +124,63 @@ protected EPersonRest createAndReturn(Context context)
+            throw new DSpaceBadRequestException("The token given already contains an email address that resolves" +
+                                                "to an eperson");
+        }
+        String emailFromJson = epersonRest.getEmail();
+        if (StringUtils.isNotBlank(emailFromJson)) {
+            if (!StringUtils.equalsIgnoreCase(registrationData.getEmail(), emailFromJson)) {
+                throw new DSpaceBadRequestException("The email resulting from the token does not match the email given"
+                                                        + " in the json body. Email from token: " +
+                                                    registrationData.getEmail() + " email from the json body: "
+                                                    + emailFromJson);
+            }
+        }
+        if (epersonRest.isSelfRegistered() != null && !epersonRest.isSelfRegistered()) {
+            throw new DSpaceBadRequestException("The self registered property cannot be set to false using this method"
+                                                    + " with a token");
+        }
+        checkRequiredProperties(epersonRest);
+        if (!AuthorizeUtil.authorizeNewAccountRegistration(context, requestService
+            .getCurrentRequest().getHttpServletRequest())) {
+            throw new DSpaceBadRequestException(
+                "Registration is disabled, you are not authorized to create a new Authorization");
+        }
+        // We'll turn off authorisation system because this call isn't admin based as it's token based
+        context.turnOffAuthorisationSystem();
+        EPerson ePerson = createEPersonFromRestObject(context, epersonRest);
+        context.restoreAuthSystemState();
+        // Restoring authorisation state right after the creation call
+        accountService.deleteToken(context, token);
+        return converter.toRest(ePerson, utils.obtainProjection());
+    }
 
-        return converter.toRest(eperson, utils.obtainProjection());
+    private void checkRequiredProperties(EPersonRest epersonRest) {
+        MetadataRest metadataRest = epersonRest.getMetadata();
+        if (metadataRest != null) {
+            List<MetadataValueRest> epersonFirstName = metadataRest.getMap().get("eperson.firstname");
+            List<MetadataValueRest> epersonLastName = metadataRest.getMap().get("eperson.lastname");
+            if (epersonFirstName == null || epersonLastName == null ||
+                epersonFirstName.isEmpty() || epersonLastName.isEmpty()) {
+                throw new DSpaceBadRequestException("The eperson.firstname and eperson.lastname values need to be " +
+                                                    "filled in");
+            }
+        }
+        String password = epersonRest.getPassword();
+        if (!accountService.verifyPasswordStructure(password)) {
+            throw new DSpaceBadRequestException("the password cannot be left blank");
```

Comment:
This error should say the password is invalid.  It's not necessarily blank, as it's also invalid if it's < 6 characters (according to the verifyPasswordStructure() method)

Reply:
This has been adjusted